src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,1000 @@
+      _igvn->register_new_node_with_optimizer(new_ptr, ptr);
+      vn->set_req(MemNode::Address, new_ptr);
+    }
+  } else if (vn->Opcode() == Op_PopulateIndex) {
+    // 2) For populate index, update start index for non-zero mask offset
+    if (level_offset != 0) {
+      int v_stride = vector_size_in_bytes / lane_size * _cl->stride_con();
+      Node* idx_off = _igvn->intcon(v_stride * level_offset);
+      Node* new_base = new AddINode(vn->in(1), idx_off);
+      _igvn->register_new_node_with_optimizer(new_base, vn->in(1));
+      vn->set_req(1, new_base);
+    }
+  }
+}
+
+// Helper function for duplicating a subgraph of nodes
+Node_List* VectorMaskedLoop::clone_node_list(const Node_List* list) {
+  assert(list != nullptr && list->size() > 0, "Should not be empty");
+  uint size = list->size();
+  Node_List* new_list = new Node_List(_arena, size);
+  Node_List* clone_map = new Node_List(_arena, size);
+  // Clone each node in the list
+  for (uint i = 0; i < size; i++) {
+    Node* old = list->at(i);
+    Node* new_node = old->clone();
+    clone_map->map(old->_idx, new_node);
+    _igvn->register_new_node_with_optimizer(new_node, old);
+    VectorNode::trace_new_vector(new_node, "VectorMasked");
+    new_list->push(new_node);
+  }
+  // Re-connect input edges to the cloned node
+  for (uint i = 0; i < size; i++) {
+    Node* new_node = new_list->at(i);
+    for (uint j = 0; j < new_node->req(); j++) {
+      Node* in = new_node->in(j);
+      if (in != nullptr && in->_idx < clone_map->max()) {
+        Node* new_in = clone_map->at(in->_idx);
+        if (new_in != nullptr) {
+          new_node->set_req(j, new_in);
+        }
+      }
+    }
+  }
+  return new_list;
+}
+
+// Entry function of actual vector mask transformation
+void VectorMaskedLoop::transform_loop(const TypeVectMask* t_vmask) {
+  // Create a tree of vector masks for different vector element sizes
+  Node_List* vmask_tree = create_vmask_tree(t_vmask);
+  Node* root_vmask = vmask_tree->at(1);
+
+  // Replace vectorization candidate nodes to vector nodes. For now we only
+  // generate a single vector node per scalar node. And that the duplication
+  // afterwards makes sure that all scalar nodes are "widened" to the same
+  // number of elements. The smalles type using a single vector, larger types
+  // using multiple (duplicated) vectors per scalar node.
+  Node_List* s2v_map = replace_scalar_ops(root_vmask);
+
+  // Duplicate and adjust vector operations with larger vector element sizes
+  // which need multiple vectors to process
+  int small = _size_stats.smallest_size();
+  int large = _size_stats.largest_size();
+  for (int lane_size = small * 2; lane_size <= large; lane_size *= 2) {
+    if (_size_stats.count_size(lane_size) > 0) {
+      duplicate_vector_ops(vmask_tree, s2v_map, lane_size);
+    }
+  }
+
+  // Update loop increment/decrement to the vector mask true count
+  Node* true_cnt = new VectorMaskTrueCountNode(root_vmask, TypeInt::INT);
```

Comment:
This seems expensive to have to use inside the loop. Is there a way we could move this outside the loop? Because if we do take the backedge then we know that we have to take the full `stride`, right?

Reply:
I guess you would have to separate out the loop-internal uses and the outside uses of the `incr`. The inside uses would use the `stride` (or is there an exception?) and the outside ones could use the `VectorMaskTrueCountNode`.

Doing something like that could have better performance.